scripts/libraries/mjpeg.py

```diff
@@ -0,0 +1,131 @@
+
+    def set_extra_headers(self, extra_headers):  # public
+        self._extra_headers = extra_headers
+
+    def _send_http_response(self, code, name, extra=""):  # private
+        self._tcp_socket.send("HTTP/1.0 %d %s\r\n%s\r\n" % (code, name, extra))
+
+    def _send_http_response_ok(self, extra=""):  # private
+        self._send_http_response(200, "OK", extra)
+
+    def _parse_mjpeg_request(self, data):  # private
+        s = data.decode().splitlines()
+        if s and len(s) >= 1:
+            line0 = s[0].split(" ")
+            request = line0[0]
+            self._pathname = re.sub("(http://[a-zA-Z0-9\-\.]+(:\d+)?(/)?)", "/", line0[1])
+            if self._pathname != "/" and self._pathname.endswith("/"):
+                self._pathname = self._pathname[:-1]
+            if line0[2] == "HTTP/1.0" or line0[2] == "HTTP/1.1":
+                if request == "GET":
+                    self._playing = True
+                    self._send_http_response_ok(
+                        "Server: OpenMV Cam\r\n"
+                        "Content-Type: multipart/x-mixed-replace;boundary=" + self._boundary + "\r\n"
+                        "Connection: close\r\n"
+                        "Cache-Control: no-cache, no-store, must-revalidate\r\n"
+                        "Pragma: no-cache\r\n"
+                        "Expires: 0\r\n"
+                        + ("\r\n".join(self._extra_headers) if len(self._extra_headers) else "")
+                    )
+                    if self._setup_cb:
+                        self._setup_cb(self._pathname)
+                    return
+                else:
+                    self._send_http_response(501, "Not Implemented")
+                    return
+        self._send_http_response(400, "Bad Request")
+
+    def _send_mjpeg(self, image_callback, quality):  # private
+        try:
+            self._tcp_socket.settimeout(5)
+            img = image_callback(self._pathname).to_jpeg(quality=quality)
+            mjpeg_header = (
+                "\r\n--" + self._boundary + "\r\n"
+                "Content-Type: image/jpeg\r\n"
+                "Content-Length:" + str(img.size()) + "\r\n\r\n"
+            )
+            self._tcp_socket.sendall(mjpeg_header)
+            self._tcp_socket.sendall(img)
+        except OSError:
+            self._close_tcp_socket()
+
+    def stream(self, image_callback, quality=90):  # public
+        while True:
+            if self._valid_tcp_socket():
+                try:
+                    self._tcp_socket.settimeout(0.01)
+                    try:
+                        data = self._tcp_socket.recv(1400)
+                        if data and len(data):
+                            self._parse_mjpeg_request(data)
+                        else:
+                            raise OSError
+                    except OSError as e:
+                        if e.errno != errno.EAGAIN and e.errno != errno.ETIMEDOUT:
+                            raise e
```

Comment:
Why the nested exception ? It's already in try/except block.

Reply:
It's what I had to do to get the rtsp server working reliably. Will switch to asyncio though which should get rid of this.